torch_xla/csrc/runtime/stablehlo_composite_helper.cc

```diff
@@ -190,35 +190,64 @@ class BuildStableHLOCompositePass : public mlir::OperationPass<mlir::ModuleOp> {
     return metadata;
   }
 
+  mlir::FailureOr<mlir::Attribute> BuildAttrFromJson(mlir::OpBuilder& builder,
+                                                     const json& json_value) {
+    switch (json_value.type()) {
+      case json::value_t::number_integer:
+      case json::value_t::number_unsigned:
+        return builder.getI64IntegerAttr(json_value.template get<int64_t>());
+      case json::value_t::number_float:
+        return builder.getF32FloatAttr(json_value.template get<float>());
+      case json::value_t::boolean:
+        return builder.getBoolAttr(json_value.template get<bool>());
+      case json::value_t::string:
+        return builder.getStringAttr(json_value.template get<std::string>());
+      case json::value_t::array: {
+        // Check if all the elements in the json array have the same type.
+        std::optional<json::value_t> common_el_type;
+        for (auto& el : json_value) {
+          auto el_type = el.type();
+          if (el_type == json::value_t::number_unsigned) {
+            el_type = json::value_t::number_integer;
+          }
+          if (common_el_type.has_value() && common_el_type.value() != el_type) {
+            return mlir::failure();
+          }
+          common_el_type = el_type;
+        }
+        if (!common_el_type.has_value()) {
+          return builder.getArrayAttr({});
+        }
```

Comment:
Control flow here was difficult to understand. How about:

```
if (json_value.empty()) return builder.getArrayAttr({});
auto headType = json_value[0].type();
bool is_homogeneous = llvm::all_of(json_value,
  [&](auto & el){ return el.type() == headType; });
if (!is_homogeneous) emitError(builder.getUnknownLocation(), ""invalid JSON to MLIR, arrays must be homogeneous");
switch (headType.value()) {
  // ....
```

Reply:
Also I'm on board with heterogeneous support, but I think it's not needed yet. Can be accomplished via a recursive call to this function.